contrib/mesos/pkg/executor/executor.go

```diff
@@ -447,308 +423,145 @@ func (k *Executor) attemptSuicide(driver bindings.ExecutorDriver, abort <-chan s
-}
 
-func (k *Executor) _launchTask(driver bindings.ExecutorDriver, taskId, podFullName string, psf podStatusFunc, expired <-chan time.Time) {
-	getMarshalledInfo := func() (data []byte, cancel bool) {
-		// potentially long call..
-		if podStatus, err := psf(); err == nil && podStatus != nil {
-			select {
-			case <-expired:
-				cancel = true
-			default:
-				k.lock.Lock()
-				defer k.lock.Unlock()
-				if _, found := k.tasks[taskId]; !found {
-					// don't bother with the pod status if the task is already gone
-					cancel = true
-					break
-				} else if podStatus.Phase != api.PodRunning {
-					// avoid sending back a running status before it's really running
-					break
-				}
-				log.V(2).Infof("Found pod status: '%v'", podStatus)
-				result := api.PodStatusResult{
-					ObjectMeta: api.ObjectMeta{
-						Name:     podFullName,
-						SelfLink: "/podstatusresult",
-					},
-					Status: *podStatus,
-				}
-				if data, err = json.Marshal(result); err != nil {
-					log.Errorf("failed to marshal pod status result: %v", err)
+	// within the launch timeout window we should see a pod-task update via k.registry.
+	// if we see a Running update then we need to generate a TASK_RUNNING status update for mesos.
+	handlerFinished := make(chan struct{})
+	expired := make(chan struct{})
+	go func() {
+		defer close(expired)
+		<-launchTimer.C
+	}()
+	handler := podWatchHandler{
+		action: func(pod *RegisteredPod) (bool, error) {
+			if pod == nil {
+				// expired!!
+				select {
+				case <-handlerFinished:
+					// noop, someone else (maybe us) has already informed mesos
+					// about the status of the task
+				default:
+					// launch timeout expired
+					k.killPodTask(driver, task.TaskId.GetValue())
 				}
+				return true, nil
 			}
-		}
-		return
-	}
 
-waitForRunningPod:
-	for {
-		select {
-		case <-expired:
-			log.Warningf("Launch expired grace period of '%v'", k.launchGracePeriod)
-			break waitForRunningPod
-		case <-time.After(containerPollTime):
-			if data, cancel := getMarshalledInfo(); cancel {
-				break waitForRunningPod
-			} else if data == nil {
-				continue waitForRunningPod
-			} else {
-				k.lock.Lock()
-				defer k.lock.Unlock()
-				task, found := k.tasks[taskId]
-				if !found {
-					goto reportLost
-				}
-
-				statusUpdate := &mesos.TaskStatus{
-					TaskId:  mutil.NewTaskID(taskId),
-					State:   mesos.TaskState_TASK_RUNNING.Enum(),
-					Message: proto.String(fmt.Sprintf("pod-running:%s", podFullName)),
-					Data:    data,
-				}
+			if pod.event == RegistrationEventDeleted {
+				defer launchTimer.Stop()
+				defer close(handlerFinished)
 
-				k.sendStatus(driver, statusUpdate)
-				task.launchTimer.Stop()
+				// we're done monitoring because pod has been deleted
+				return true, nil
+			}
 
-				// continue to monitor the health of the pod
-				go k.__launchTask(driver, taskId, podFullName, psf)
-				return
+			if !podTaskRunning(pod) {
```

Comment:
probably want to filter on pod.event here: only worth proceeding if it's an update event and the pod is running